### DISCUSSION TITLE
How do I handle precedence between truly ambiguous symbols

### DISCUSSION THREAD

**Discussion by Hubro** [tree-sitter/tree-sitter#1275] [Q&A] [answered]
Yesterday I learned, to my horror, that the language I'm trying to implement a grammar for (YANG) supports *unquoted strings*. That is, as long as a string doesn't contain any white space or any quotes, omitting the surrounding quotes is acceptable.

Obviously, as soon as I add support for this, the new "unquoted_string" symbol now conflicts with, and overrides, existing types like "range" and "date". These value types are completely ambiguous, so I want "unquoted_string" to have the absolute lowest priority.

This seems like it should be an easy fix using `prec.dynamic`, but after tearing my hair out for over an hour I can't get it to make any difference.

```javascript
conflicts: $ => [
    [$.string, $.range],
    [$.unquoted_string, $.range, $.date],
],
```

```javascript
argument: $ => choice(
    $.built_in_type,
    $.node_identifier,
    $.integer,
    $.string,
    $.unquoted_string,
    $.string_concatenation,
    $.date,
    $.range,
    $.keypath,
    $.yang_version,
    $.glob,
),
```

```javascript
unquoted_string: $ => prec.dynamic(-1, /[a-zA-Z0-9-_:.]+/),
```

```javascript
range: $ => choice(
    $.unquoted_range,
    $.quoted_range,
),

unquoted_range: $ => /\d+\.\.\d+/,
quoted_range: $ => /"\d+\.\.\d+"/,
```

"unquoted_string" still overrides anything that is ambiguous with it:

![image](https://user-images.githubusercontent.com/597206/126025261-42f35161-8cb4-4f9f-b71e-6c7f0d4e3ddc.png)

If I just delete the "unquoted_string" option from "argument", the tests succeed again.

The documentation on dynamic precedence is very sparse and has no examples, but I've looked through many other grammars though and it looks like I'm using it correctly.

Am I missing anything?

---

Full grammar: https://github.com/Hubro/tree-sitter-yang/blob/master/grammar.js

**Comment by Hubro** [comment]
So basically my problem sums down to:

- An "argument" in my grammar can be a date (`2012-12-31`), a range (`1..100`) or an "unquoted string", which matches pretty much anything
- As soon as I add the "unquoted string" regex into my `choice(...)`, the parser will ***always*** go straight for it, meaning that the date and range patterns will never be attempted
- All I need to do is instruct the parser to try the "unquoted string" regex ***last***, so if an argument matches a date, it will be parsed as a date symbol

I have tried everything I can think of, but nothing I do makes any difference. This seems like it should be so trivial :confused: 

I made a small test case for the ambiguity case and debugged it:

```yang
module foo {
    range 1..100;
}
```

Here is the output:

```
new_parse
process version:0, version_count:1, state:1, row:1, col:0
lex_internal state:37, row:1, column:0
  consume character:'m'
  consume character:'o'
  consume character:'d'
  consume character:'u'
  consume character:'l'
  consume character:'e'
lexed_lookahead sym:module, size:6
shift state:36
process version:0, version_count:1, state:36, row:1, col:6
lex_internal state:12, row:1, column:6
  consume character:' '
lexed_lookahead sym:_sep, size:1
shift_extra
process version:0, version_count:1, state:36, row:1, col:7
lex_internal state:12, row:1, column:7
  consume character:'f'
  consume character:'o'
  consume character:'o'
lexed_lookahead sym:extension_keyword_token1, size:3
shift state:51
process version:0, version_count:1, state:51, row:1, col:10
lex_internal state:37, row:1, column:10
  consume character:' '
lexed_lookahead sym:_sep, size:1
shift_extra
process version:0, version_count:1, state:51, row:1, col:11
lex_internal state:37, row:1, column:11
  consume character:'{'
lexed_lookahead sym:{, size:1
reduce sym:identifier, child_count:1
shift state:5
process version:0, version_count:1, state:5, row:1, col:12
lex_internal state:11, row:1, column:12
  consume character:10
lexed_lookahead sym:_sep, size:1
shift_extra
process version:0, version_count:1, state:5, row:2, col:0
lex_internal state:11, row:2, column:0
  consume character:' '
lexed_lookahead sym:_sep, size:1
shift_extra
process version:0, version_count:1, state:5, row:2, col:1
lex_internal state:11, row:2, column:1
  consume character:' '
lexed_lookahead sym:_sep, size:1
shift_extra
process version:0, version_count:1, state:5, row:2, col:2
lex_internal state:11, row:2, column:2
  consume character:' '
lexed_lookahead sym:_sep, size:1
shift_extra
process version:0, version_count:1, state:5, row:2, col:3
lex_internal state:11, row:2, column:3
  consume character:' '
lexed_lookahead sym:_sep, size:1
shift_extra
process version:0, version_count:1, state:5, row:2, col:4
lex_internal state:11, row:2, column:4
  consume character:'r'
  consume character:'a'
  consume character:'n'
  consume character:'g'
  consume character:'e'
lexed_lookahead sym:range, size:5
shift state:45
process version:0, version_count:1, state:45, row:2, col:9
lex_internal state:37, row:2, column:9
  consume character:' '
lexed_lookahead sym:_sep, size:1
reduce sym:statement_keyword, child_count:1
shift state:15
process version:0, version_count:1, state:15, row:2, col:10
lex_internal state:1, row:2, column:10
  consume character:'1'
  consume character:'.'
  consume character:'.'
  consume character:'1'
  consume character:'0'
  consume character:'0'
lexed_lookahead sym:unquoted_string, size:6
shift state:32
process version:0, version_count:1, state:32, row:2, col:16
lex_internal state:37, row:2, column:16
  consume character:';'
lexed_lookahead sym:;, size:1
reduce sym:argument, child_count:1
shift state:11
process version:0, version_count:1, state:11, row:2, col:17
lex_internal state:11, row:2, column:17
  consume character:10
lexed_lookahead sym:_sep, size:1
shift_extra
process version:0, version_count:1, state:11, row:3, col:0
lex_internal state:11, row:3, column:0
  consume character:'}'
lexed_lookahead sym:}, size:1
reduce sym:statement, child_count:4
shift state:54
process version:0, version_count:1, state:54, row:3, col:1
lex_internal state:37, row:3, column:1
  consume character:10
lexed_lookahead sym:_sep, size:1
shift_extra
process version:0, version_count:1, state:54, row:4, col:0
lex_internal state:37, row:4, column:0
lexed_lookahead sym:end, size:0
reduce sym:block, child_count:3
reduce sym:module, child_count:3
reduce sym:yang, child_count:1
accept
done
(yang [0, 0] - [3, 0]
  (module [0, 0] - [2, 1]
    module_name: (identifier [0, 7] - [0, 10])
    module_block: (block [0, 11] - [2, 1]
      (statement [1, 4] - [1, 17]
        (statement_keyword [1, 4] - [1, 9])
        (argument [1, 10] - [1, 16]
          (unquoted_string [1, 10] - [1, 16]))))))
```

([HTML graph version](https://gist.githubusercontent.com/Hubro/ad8ef80b838c6680738c4291a292d82e/raw/002d1c7949c465a62d7ccf94bdece06c6306df16/log.html))

Any input on this would be greatly appreciated.

**Reply by Hubro** [reply]
I've boiled my problem down into a minimal example:

```javascript
module.exports = grammar({
    name: "ambiguitytest",

    conflicts: $ => [
        [$.unquoted_string, $.unquoted_range],
    ],

    rules: {
        statement: $ => choice(
            $.unquoted_range,
            $.unquoted_string,
        ),

        unquoted_range: $ => prec.dynamic(-1, /\d+\.\.\d+/),
        unquoted_string: $ => prec.dynamic(10, /[a-zA-Z0-9-_.:^]+/),
    },
});
```

When run with `test.txt`:

```
0..100
```

The statement is ***always*** parsed as `unquoted_range`. No matter what I try, I can't get it to be parsed as `unquoted_string`. `prec.dynamic` doesn't do anything, plain `prec` doesn't do anything, the order of the symbols inside `choice(...)` doesn't matter. If I comment away `$.unquoted_range`, the symbol becomes parsed as an `unquoted_string` instead, so clearly both symbols apply.

I don't understand why I can't affect the priority of the symbols in any way. Surely it must be possible, so what am I missing?

---

Oh my, by random experimentation I noticed that the order in which the symbol is defined in the `rules` object *does* actually affect the precedence... In other words, if I swap these two lines:

```javascript
{
    // ...
    unquoted_range: $ => prec.dynamic(-1, /\d+\.\.\d+/),
    unquoted_string: $ => prec.dynamic(10, /[a-zA-Z0-9-_.:^]+/),
    // ...
}
```

So the string definition comes first, the statement is parsed as a `unquoted_string` instead... :thinking: 

This feels like an implementation detail that I shouldn't rely on, though. I would really like to understand how I can "properly" affect the precedence.

**Reply by ahelwer** [reply]
Doesn't address your issue generally but precedence rules like the order definition principle you discovered are documented [here](https://tree-sitter.github.io/tree-sitter/creating-parsers#conflicting-tokens).

**Reply by Hubro** [reply]
> Doesn't address your issue generally but precedence rules like the order definition principle you discovered are documented [here](https://tree-sitter.github.io/tree-sitter/creating-parsers#conflicting-tokens).

Ah right, yes I did read that a while ago... So it seems like the precedence functions should have higher priority than the rule order. I must not be using them correctly :thinking: I feel like I've tried everything though.

**Reply by ahelwer** [reply]
@Hubro I'll play around with your grammar in a bit but one thing that comes to mind is an issue I encountered a few months back where I couldn't get GLR parsing to trigger at all: https://github.com/tree-sitter/tree-sitter/issues/1061

**Comment by ahlinc** [comment, marked as answer]
@Hubro your issue is easily fixable by moving the `unquoted_string` definition bellow `unquoted_range` and `date` definitions.
See full `git diff` bellow. **Update:** _For another option see my second comment to this post._

_Before:_

![image](https://user-images.githubusercontent.com/14666676/126069959-a972dec3-42e8-4303-a2f8-f116291dc5c4.png)

_After:_

![image](https://user-images.githubusercontent.com/14666676/126069968-ed9977fa-c0b6-4eee-b9d8-d62294c20c0e.png)

<details><summary><i>git diff:</i></summary><p>

```diff
diff --git i/grammar.js w/grammar.js
index fc31635..977e2ba 100644
--- i/grammar.js
+++ w/grammar.js
@@ -21,7 +21,6 @@ const yang_grammar = grammar({

     conflicts: $ => [
         [$.string, $.range],
-        [$.unquoted_string, $.range, $.date],
     ],

     rules: {
@@ -141,11 +140,6 @@ const yang_grammar = grammar({
             )
         ),

-        // Unquoted strings are not explained in the ABNF grammar, so we're
-        // going to assume it can be any identifier character plus a few more
-        // common symbols.
-        unquoted_string: $ => prec.dynamic(-1, /[a-zA-Z0-9-_:.]+/),
-
         string_concatenation: $ => seq(
             repeat1(seq($.string, alias('+', $.plus_symbol))),
             $.string
@@ -178,6 +172,11 @@ const yang_grammar = grammar({
         unquoted_range: $ => /\d+\.\.\d+/,
         quoted_range: $ => /"\d+\.\.\d+"/,

+        // Unquoted strings are not explained in the ABNF grammar, so we're
+        // going to assume it can be any identifier character plus a few more
+        // common symbols.
+        unquoted_string: $ => /[a-zA-Z0-9-_:.]+/,
+
         keypath: $ => token(
             choice(
                 absolute_keypath(),
```

</p></details>

**Reply by ahlinc** [reply]
Your issue isn't fixable with the **GLR** algorithm and the `conflicts` statement because the `conflicts` statement helps to resolve conflicts when there no possibility for the parser to make decision by looking ahead for a one token but the conflict can be solved by looking forward for several grammar tokens. In other words the `conflicts` statement helpful in resolving conflicts between grammar rules applied to a sequence of grammar tokens _(terminals)_ where such rules are ambiguous in a some step of the sequence and to solve the ambiguity there is the need for the lookahead more than on a one token _(terminal)_.
The issue that you have is the issue between single tokens that can happen in the same place and there only the tokens precedence may help or the use of regexps that don't clash with each other.
Solution suggested above corresponds to the item 5 from https://tree-sitter.github.io/tree-sitter/creating-parsers#conflicting-tokens

**Reply by ahlinc** [reply]
Another solution that corresponds to the second item from the above mentioned link is to apply `token(prec(1, ...))` for the `date` and `unquoted_range` tokens:

<details><summary><i>git diff:</i></summary><p>

```diff
diff --git i/grammar.js w/grammar.js
index fc31635..4fcba1c 100644
--- i/grammar.js
+++ w/grammar.js
@@ -21,7 +21,6 @@ const yang_grammar = grammar({

     conflicts: $ => [
         [$.string, $.range],
-        [$.unquoted_string, $.range, $.date],
     ],

     rules: {
@@ -144,7 +143,7 @@ const yang_grammar = grammar({
         // Unquoted strings are not explained in the ABNF grammar, so we're
         // going to assume it can be any identifier character plus a few more
         // common symbols.
-        unquoted_string: $ => prec.dynamic(-1, /[a-zA-Z0-9-_:.]+/),
+        unquoted_string: $ => /[a-zA-Z0-9-_:.]+/,

         string_concatenation: $ => seq(
             repeat1(seq($.string, alias('+', $.plus_symbol))),
@@ -168,14 +167,14 @@ const yang_grammar = grammar({
             )
         )),

-        date: $ => /\d{4}-\d{2}-\d{2}/,
+        date: $ => token(prec(1, /\d{4}-\d{2}-\d{2}/)),

         range: $ => choice(
             $.unquoted_range,
             $.quoted_range,
         ),

-        unquoted_range: $ => /\d+\.\.\d+/,
+        unquoted_range: $ => token(prec(1, /\d+\.\.\d+/)),
         quoted_range: $ => /"\d+\.\.\d+"/,

         keypath: $ => token(
```

</p></summary>

**Reply by Hubro** [reply]
YES! It blows my mind why this is the case, but wrapping `prec(...)` inside `token(...)` makes it work as intended :exploding_head: 

But it completely fixes my problem! Thank you!!!

Is there anything in the documentation that I've missed that explains this behavior?

**Reply by ahelwer** [reply]
@Hubro from [creating parsers#conflicting tokens](https://tree-sitter.github.io/tree-sitter/creating-parsers#conflicting-tokens) (emphasis added):
> **Lexical Precedence** - When the precedence functions described above are used *within the token function*, the given precedence values serve as instructions to the lexer. If there are two valid tokens that match the characters at a given position in the document, Tree-sitter will select the one with the higher precedence.

So I guess if you want to use precedence at the lexing level rather than the parsing level you have to wrap it in a `token`.

**Reply by Hubro** [reply]
Crap, I've literally read that paragraph probably 6 times the last few days, and I completely overlooked that it said "within the token function". :man_facepalming: 

I'm not really clear on the difference between the "parsing level" and the "lexing level". Guess I should probably read up on some theory.

**Reply by ahlinc** [reply]
> Crap, I've literally read that paragraph probably 6 times the last few days, and I completely overlooked that it said "within the token function".

Don't worry! You're not alone :smile: My reaction at the first time was the same.

**Reply by ahelwer** [reply]
@Hubro no worries, I only just realized what it meant as a result of this question (I'm still a bit hazy on the purpose of `token` in general). The difference between lexing and parsing is not too tricky to understand, they're just very integrated in tree-sitter. Basically lexing (or scanning) is the process of taking the raw text and turning it into tokens, like how `2 + 3 - 5` is turned into the lexeme sequence `2`, `+`, `3`, `-`, `5`, then into the token sequence `(number) (plus) (number) (minus) (number)`. Then parsing can take that sequence of tokens and match it with a set of rules to turn it into a parse tree like `(minus (plus (number) (number)) (number))`. With tree-sitter you can go quite a while without having to care about lexing at all, as long as your needs are simple. Once you start caring about [regex lookahead](https://github.com/tree-sitter/tree-sitter/discussions/1252) or [overriding tree-sitter's contextual lexing behavior](https://github.com/tree-sitter/tree-sitter/discussions/1257) or [parsing context-sensitive expressions](https://github.com/tree-sitter/tree-sitter/discussions/1215#discussioncomment-938407) you start thinking about moving that behavior into an external scanner rather than specifying it in your grammar file.
